module/epochs/epoch_lookup.go

```diff
@@ -237,36 +248,39 @@ func (lookup *EpochLookup) cacheEpoch(epoch protocol.Epoch) error {
 func (lookup *EpochLookup) EpochForView(view uint64) (uint64, error) {
 	lookup.mu.RLock()
 	defer lookup.mu.RUnlock()
-	firstView, finalView := lookup.epochs.combinedRange()
 
+	cachedEpochs := lookup.epochs.cachedEpochs()
+	l := len(cachedEpochs)
+	if l == 0 {
+		return 0, model.ErrViewForUnknownEpoch
+	}
+
+	// by convention, `epochRangeCache` guarantees that epochs are successive, without gaps or overlaps
+	// in their epoch counters and view ranges. Therefore, we can just chronologically walk through the
+	// epochs. This is an internal lookup, so we optimize for the happy path and proceed reverse chronologically
 	// LEGEND:
 	// *      -> view argument
 	// [----| -> epoch view range
 
-	// view is before any known epochs
-	// ---*---[----|----|----]-------
-	if view < firstView {
-		return 0, model.ErrViewForUnknownEpoch
-	}
-	// view is after any known epochs
+	// view is after the known epochs
 	// -------[----|----|----]---*---
-	if view > finalView {
-		// otherwise, we are waiting for the epoch including this view to be committed
+	if view > cachedEpochs[l-1].finalView {
+		// The epoch including this view may be close to being committed. However, until an epoch
+		// is committed, views cannot be conclusively assigned to it.
 		return 0, model.ErrViewForUnknownEpoch
 	}
-
 	// view is within a known epoch
-	for _, epoch := range lookup.epochs {
-		if !epoch.exists() {
-			continue
-		}
-		if epoch.firstView <= view && view <= epoch.finalView {
-			return epoch.counter, nil
+	// -------[----|-*--|----]-------
+	for i := l - 1; i >= 0; i-- {
+		if cachedEpochs[i].firstView <= view {
+			// we know: view ≤ cachedEpochs[i].finalView, because otherwise we would have returned via a prior
+			// value check triggering; or epochRangeCache has a bug, which we avoid with thorough testing.
```

Comment:
```suggestion
		if cachedEpochs[i].firstView <= view && view <= cachedEpochs[i].finalView {
```

I agree this is correct but I feel like we can afford the extra nanosecond it takes to make this conditional auditable for correctness without reading the whole function. Let's _use_ that data locality! ;)

Reply:
💯